File: bootstrap/tokenizer.c

// tokenizer: group into tokens based on whitespace
#include <stddef.h>;
#include <string.h>;
#include <stdbool.h>;
#include "tokenizer.h";

// TODO add assert and error functions / macros

bool addToken(struct token *tokens, char *tokenStartIndex, char *tokenEndIndex,
              unsigned int lineNum, unsigned int colNum)
{
    struct token *token = malloc(sizeof(token));

    token->tokenStartIndex = tokenStartIndex;
    token->tokenEndIndex = tokenEndIndex;
    token->lineNum = lineNum;
    token->colNum = colNum;

    // TODO now add token to tokensList
    return false;
}

bool pushToken(struct token *tokens, char *tokenStartIndex, char *tokenEndIndex,
               unsigned int lineNum, unsigned int colNum)
{
    const currPartialTokenLen = tokenEndIndex - tokenStartIndex;
    const isCurrPartialTokenEmpty = tokenStartIndex == tokenEndIndex;

    if (!isCurrPartialTokenEmpty)
        return addToken(tokens, tokenStartIndex, tokenEndIndex,
                        lineNum, colNum - currPartialTokenLen);
}

bool pushCharToken(struct token *tokens, char *tokenStartIndex, char *tokenEndIndex,
                   unsigned int lineNum, unsigned int colNum)
{
    assert(tokenStartIndex + 1 == tokenEndIndex,
           "Internal error: tokenStartIndex must equal tokenEndIndex");

    // -1 for length, + 1 for col num increment
    // after each iteration
    return addToken(tokens, tokenStartIndex, tokenEndIndex,
                    lineNum, colNum - 1 + 1);
}

// TODO add return for token struct
// TODO we're going to need a linkedlist for tokens
int tokenize(char *str)
{
    // TODO
    int tokens = malloc(1);

    unsigned int lineNum = 1;
    unsigned int colNum = 1;
    char *currPartialToken = "";

    // make sure all tokens are pushed out
    str += '\n';
    
    const len = strlen(str);

    for (unsigned int i = 0; i < len; i++)
    {
        const currChar = str[i];

        switch (currChar)
        {
        case '\n':
            pushToken();
            pushChar(currChar);

            // update line and column nums
            lineNum++;
            colNum = 0;
            currPartialToken = "";

            break;
        case ' ':
        case '\t':
            pushToken();
            currPartialToken = "";
            break;
        CASE_NON_ALPHANUM:
            pushToken();
            pushChar(currChar);
            currPartialToken = "";

            break;
        CASE_ALPHANUM:
            currPartialToken += currChar;
            break;
        default:
            error("Invalid char '%c' [%i:%i]", currChar, lineNum, colNum);
        }

        colNum++;
    }

    return tokens;
}

// lexer: uses context to assign meaning to tokens

File: bootstrap/tokenizer.js

// tokenizer: group into tokens based on whitespace
const tokenize = (str) => {
    const tokens = [];

    let lineNumber = 1;
    let columnNumber = 1;
    let currPartialToken = "";

    const pushToken = () => currPartialToken && tokens.push({
        token: currPartialToken,
        lineNumber: lineNumber,
        columnNumber: columnNumber - currPartialToken.length,
    })

    const pushChar = (currChar) => tokens.push({
        token: currChar,
        lineNumber: lineNumber,
        // -1 for length, + 1 for col num increment
        // after each iteration
        columnNumber: columnNumber - 1 + 1,
    })

    // make sure all tokens are pushed out
    str += "\n";
    let prevChar = -1;
    let nextChar = -1;
    for (let i = 0; i < str.length; i++) {
        const currChar = str.charAt(i);
        nextChar = str.charAt(i + 1) || -1;

        switch (currChar) {
            case "\n":
                pushToken();
                pushChar(currChar);

                // update line and column nums
                lineNumber++;
                columnNumber = 0;
                currPartialToken = "";

                break;
            case " ":
            case "\t":
                pushToken();
                currPartialToken = "";
                break;

            // TODO consider splitting out ()[]{} into separate
            // cases since they should separate from the others
            // as separate tokens
            case "!": case "\"": case "#": case "$": case "%":
            case "&": case "'": case "(": case ")": case "*":
            case "+": case ",": case "-": case ".": case "/":
            case ":": case ";": case "<": case "=": case ">":
            case "?": case "@": case "[": case "\\": case "]":
            case "^": case "_": case "`": case "{": case "|":
            case "}":
                // prevChar or nextChar is a special char ;
                // this means that the token will be continued
                // so we just keep adding to the token
                if (prevChar == "!" || prevChar == "\"" || prevChar == "#" ||
                    prevChar == "$" || prevChar == "%" || prevChar == "&" ||
                    prevChar == "'" || prevChar == "(" || prevChar == ")" ||
                    prevChar == "*" || prevChar == "+" || prevChar == "," ||
                    prevChar == "-" || prevChar == "." || prevChar == "/" ||
                    prevChar == ":" || prevChar == ";" || prevChar == "<" ||
                    prevChar == "=" || prevChar == ">" || prevChar == "?" ||
                    prevChar == "@" || prevChar == "[" || prevChar == "\\" ||
                    prevChar == "]" || prevChar == "^" || prevChar == "_" ||
                    prevChar == "`" || prevChar == "{" || prevChar == "|" ||
                    prevChar == "}" ||

                    nextChar == "!" || nextChar == "\"" || nextChar == "#" ||
                    nextChar == "$" || nextChar == "%" || nextChar == "&" ||
                    nextChar == "'" || nextChar == "(" || nextChar == ")" ||
                    nextChar == "*" || nextChar == "+" || nextChar == "," ||
                    nextChar == "-" || nextChar == "." || nextChar == "/" ||
                    nextChar == ":" || nextChar == ";" || nextChar == "<" ||
                    nextChar == "=" || nextChar == ">" || nextChar == "?" ||
                    nextChar == "@" || nextChar == "[" || nextChar == "\\" ||
                    nextChar == "]" || nextChar == "^" || nextChar == "_" ||
                    nextChar == "`" || nextChar == "{" || nextChar == "|" ||
                    nextChar == "}") {
                    currPartialToken += currChar;
                }
                else {
                    pushToken();
                    pushChar(currChar);
                    currPartialToken = "";
                }

                break;
            // nums
            case "0": case "1": case "2": case "3": case "4":
            case "5": case "6": case "7": case "8": case "9":

            // uppercase letters
            case "A": case "B": case "C": case "D": case "E":
            case "F": case "G": case "H": case "I": case "J":
            case "K": case "L": case "M": case "N": case "O":
            case "P": case "Q": case "R": case "S": case "T":
            case "U": case "V": case "W": case "X": case "Y":
            case "Z":

            // lowercase letters
            case "a": case "b": case "c": case "d": case "e":
            case "f": case "g": case "h": case "i": case "j":
            case "k": case "l": case "m": case "n": case "o":
            case "p": case "q": case "r": case "s": case "t":
            case "u": case "v": case "w": case "x": case "y":
            case "z":
                currPartialToken += currChar;
                break;
            default:
                throw new Error(`Invalid char '${currChar}' [${lineNumber}:${columnNumber}]`);
        }

        prevChar = currChar;

        columnNumber++;

    }

    return tokens
}
